src/transformer/lazy_string.hpp

```diff
@@ -0,0 +1,129 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#pragma once
+
+#include <cstdlib>
+#include <cstring>
+#include <ddwaf.h>
+#include <stdexcept>
+#include <string_view>
+
+namespace ddwaf {
+
+class lazy_string {
+public:
+    explicit lazy_string(std::string_view original)
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+        : buffer_(const_cast<char *>(original.data())), length_(original.length())
+    {
+        if (buffer_ == nullptr) {
+            throw std::runtime_error{"lazy_string initialised with nullptr"};
+        }
+    }
+
+    lazy_string(const lazy_string &) = delete;
+    lazy_string &operator=(const lazy_string &) = delete;
+    lazy_string(lazy_string &&other) = delete;
+    lazy_string &operator=(lazy_string &&other) = delete;
+
+    ~lazy_string()
+    {
+        if (modified_) {
+            // NOLINTNEXTLINE(hicpp-no-malloc,cppcoreguidelines-no-malloc)
+            free(buffer_);
+        }
+    }
+
+    [[nodiscard]] constexpr const char &at(std::size_t idx) const { return buffer_[idx]; }
+
+    char &operator[](std::size_t idx)
```

Comment:
nitpicking but it seems you use `at` much more often, so maybe you could use `operator[]` for that one instead

Reply:
For consistency with `std::string`, I prefer to use `[]` to get a reference.